### ISSUE TITLE
Has freezing parameters behaviour changed? None values are not allowed?

### ISSUE THREAD

**Issue by cottrell** [patrick-kidger/equinox#908]
I see this error 

```
In previous releases of JAX, flatten-up-to used to consider None to be a tree-prefix of non-None values. To obtain the previous behavior, you can usually write:
  jax.tree.map(lambda x, y: None if x is None else f(x, y), a, b, is_leaf=lambda x: x is None)
```

But it is not obvious what "f" is.

The freezing example does not produce this error so something in the example is perhaps incomplete.

I don't have much time now but will try to get a complete example. I've done these filtering things several times and always feel like the docs are not quite stable or complete.  What I will aim to re-produce in a minimal example is this:
- freezing params should involve leaking the structure of the module into the param freezing routing. Freezing shoudl be done based on names, tags, class names etc. Not nth thing to the right etc.
- freezing params should be done on modules, not just arrays.
- freezing params shoudl be illustrated as a **delta** to non-freezed params. i.e. comment out the loss function and show the pre/post loss/grad manipulations.

**Comment by cottrell** [comment]
This is my current working example modified from https://docs.kidger.site/equinox/examples/frozen_layer/

I think either I've forgotten what the alternative approaches are or we all keep hitting on the problem that we do not want to have the model structure leak into filter mechanism. Basically, we want something like static_field in the "one true place" the model structure is defined but for that to be some kind of tag that we can condition on reliably.

I think you can maybe do this via name, class, name of module but due to the way trees vs modules work it becomes a bit complicated?

You can see below several cases, some of which do not work yet.

If it isn't obvious, the use case is dev flow where you want to avoid chasing your debug freezes every time you change something about the model. For example, I might want to freeze all things related to "covariance" (whatever that might mean) in all models I am using while I try something out. It seems like you can't catch the names of modules in the same way you catch the names of Arrays.


```python
import equinox as eqx
import jax
import jax.numpy as jnp
import jax.random as jrandom
import jax.tree_util as jtu
import optax  # https://github.com/deepmind/optax

def get_data(dataset_size, *, key):
    x = jrandom.normal(key, (dataset_size, 1))
    y = 5 * x - 2
    return x, y


def dataloader(arrays, batch_size, *, key):
    dataset_size = arrays[0].shape[0]
    assert all(array.shape[0] == dataset_size for array in arrays)
    indices = jnp.arange(dataset_size)
    while True:
        perm = jrandom.permutation(key, indices)
        (key,) = jrandom.split(key, 1)
        start = 0
        end = batch_size
        while end < dataset_size:
            batch_perm = perm[start:end]
            yield tuple(array[batch_perm] for array in arrays)
            start = end
            end = start + batch_size


def filter_by_name(model, names):
    def match_fn(node, name):
        check = name.name in names
        print(f'Checking {name.name} in {names}: {check}')
        return not check  # confusing, why "not"
    
    filter_spec = jtu.tree_map_with_path(
        lambda path, leaf: match_fn(leaf, path[-1]),
        model
    )
    return filter_spec

def freeze_all_modules_of_type(model, target_type):
    def match_fn(node, inside_target=False):
        if isinstance(node, target_type):
            print(f"Entering target module: {node.__class__.__name__}")
            inside_target = True  # Set inside_target when we enter the target type
        if isinstance(node, eqx.Module):
            # Traverse fields of the module while tracking inside_target state
            return jtu.tree_map(
                lambda subnode: match_fn(subnode, inside_target),
                node,
                is_leaf=lambda x: not isinstance(x, eqx.Module)
            )
        # If inside_target is True, freeze this leaf
        print(f"Node: {type(node)}, Inside Target: {inside_target}")
        return inside_target

    return match_fn(model)


class CustomModule(eqx.Module):
    A: eqx.Module
    B: eqx.Module
    C: eqx.Module

    def __init__(self, *, key, in_size, out_size, width_size, depth):
        keys = jrandom.split(key, 2)
        self.A = eqx.nn.MLP(in_size, out_size, width_size, depth, key=keys[0])
        self.B = eqx.nn.MLP(in_size, out_size, width_size, depth, key=keys[1])
        self.C = eqx.nn.Linear(in_size, out_size, key=keys[1])

    def __call__(self, x):
        return self.A(x) + self.B(x) + self.C(x)

def main(
    dataset_size=10000,
    batch_size=256,
    learning_rate=3e-3,
    steps=30,
    width_size=2,
    depth=1,
    seed=5678,
    frozen=None,
):
    data_key, loader_key, model_key = jrandom.split(jrandom.PRNGKey(seed), 3)
    data = get_data(dataset_size, key=data_key)
    data_iter = dataloader(data, batch_size, key=loader_key)

    model = CustomModule(key=model_key, in_size=1, out_size=1, width_size=width_size, depth=depth)

    filter_spec = None
    if frozen == 'name_bias':
        # NOTE: works
        filter_spec = filter_by_name(model, ['bias'])
    elif frozen == 'name_A':
        # NOTE: does not work
        filter_spec = filter_by_name(model, ['A'])
    elif frozen == 'class':
        # NOTE: does not work
        filter_spec = freeze_all_modules_of_type(model, eqx.nn.MLP)
    elif frozen == 'class_name':
        raise NotImplementedError('Not implemented')

    def loss_fn(model, x, y):
        pred_y = jax.vmap(model)(x)
        return jnp.mean((y - pred_y) ** 2)

    def loss_split(diff_model, static_model, x, y):
        model = eqx.combine(diff_model, static_model)
        return loss_fn(model, x, y)

    @eqx.filter_jit
    def make_step(model, x, y, opt_state):
        if frozen:
            diff_model, static_model = eqx.partition(model, filter_spec)
            grads = eqx.filter_grad(loss_split)(diff_model, static_model, x, y)
        else:
            grads = eqx.filter_grad(loss_fn)(model, x, y)
        updates, opt_state = optim.update(grads, opt_state)
        model = eqx.apply_updates(model, updates)
        return model, opt_state

    original_model = model
    optim = optax.sgd(learning_rate)
    opt_state = optim.init(model)
    for step, (x, y) in zip(range(steps), data_iter):
        print(
            model.A.layers[0].weight[0][0], model.A.layers[0].bias[0],
            model.B.layers[0].weight[0][0], model.B.layers[0].bias[0],
            model.C.weight[0][0], model.C.bias[0]
        )
        model, opt_state = make_step(model, x, y, opt_state)
```

**Comment by johannahaffner** [comment]
Hi David, 

I do not get any errata on 

```
equinox : 0.11.10
jax version: 0.4.36
optax version: 0.2.4
```

Does that solve your problem? 

Regarding your question about filtering arrays vs other elements:

Filtering does indeed split the pytree into dynamic and static components depending on the definition of `filter_spec`. However this does not mean that the model structure (its PyTree structure) is only defined in "the one true place". Partition creates two PyTrees of identical structure, one which has `None` leaves in place of everything static, and one which has `None` leaves in place of everything marked as dynamic.

In the [simple example](https://docs.kidger.site/equinox/all-of-equinox/#2-filtering), it just so happens that `filter_spec` divides arrays from non-arrays, but you can absolutely divide according to other criteria, as you did above.

**Comment by cottrell** [comment]
Yes, upgrading seems to fix it.

These were my various patterns for testing FWIW:

```python
import equinox as eqx
import jax
import jax.numpy as jnp
import jax.random as jrandom
import jax.tree_util as jtu
import optax  # https://github.com/deepmind/optax

def get_data(dataset_size, *, key):
    x = jrandom.normal(key, (dataset_size, 1))
    y = 5 * x - 2
    return x, y


def dataloader(arrays, batch_size, *, key):
    dataset_size = arrays[0].shape[0]
    assert all(array.shape[0] == dataset_size for array in arrays)
    indices = jnp.arange(dataset_size)
    while True:
        perm = jrandom.permutation(key, indices)
        (key,) = jrandom.split(key, 1)
        start = 0
        end = batch_size
        while end < dataset_size:
            batch_perm = perm[start:end]
            yield tuple(array[batch_perm] for array in arrays)
            start = end
            end = start + batch_size


def filter_by_name(model, names):
    def match_fn(key, value):
        check = True
        for k in key:
            if hasattr(k, 'name') and k.name in names:
                check = False
                break
        # print(f'Key: {key}, Check: {check}')
        return check
    
    return jtu.tree_map_with_path(lambda key, value: match_fn(key, value), model)


def filter_all_modules_by_type(model, target_types):
    def node_f(key, value):
        check = isinstance(value, target_types)
        # print(f'key: {key}, check: {check}')
        return not check

    def is_leaf_f(node):
        check = isinstance(node, target_types)
        return check

    return jtu.tree_map_with_path(node_f, model, is_leaf=is_leaf_f)

def filter_all_modules_by_class_name(model, target_names):
    def node_f(key, value):
        check = False
        if hasattr(value, '__class__'):
            check = value.__class__.__name__ in target_names
        # print(f'key: {key}, check: {check}')
        return not check

    def is_leaf_f(node):
        check = False
        if hasattr(node, '__class__'):
            check = node.__class__.__name__ in target_names
        return check

    return jtu.tree_map_with_path(node_f, model, is_leaf=is_leaf_f)



class CustomModule(eqx.Module):
    A: eqx.Module
    B: eqx.Module
    C: eqx.Module

    def __init__(self, *, key, in_size, out_size, width_size, depth):
        keys = jrandom.split(key, 3)
        self.A = eqx.nn.MLP(in_size, out_size, width_size, depth, key=keys[0])
        self.B = eqx.nn.MLP(in_size, out_size, width_size, depth, key=keys[1])
        self.C = eqx.nn.Linear(in_size, out_size, key=keys[2])

    def __call__(self, x):
        return self.A(x) + self.B(x) + self.C(x)



def get_model(width_size=2, depth=1, seed=5678):
    model_key = jrandom.PRNGKey(seed)
    return CustomModule(key=model_key, in_size=1, out_size=1, width_size=width_size, depth=depth)

def main(
    dataset_size=10000,
    batch_size=256,
    learning_rate=3e-3,
    steps=100,
    width_size=2,
    depth=1,
    seed=5678,
    frozen=None,
):
    data_key, loader_key, model_key = jrandom.split(jrandom.PRNGKey(seed), 3)
    data = get_data(dataset_size, key=data_key)
    data_iter = dataloader(data, batch_size, key=loader_key)

    model = CustomModule(key=model_key, in_size=1, out_size=1, width_size=width_size, depth=depth)

    filter_spec = None
    if frozen == 'name_bias':
        # NOTE: filter based on name anywhere in the path
        filter_spec = filter_by_name(model, ('bias',))
    elif frozen == 'name_A':
        # NOTE: filter based on name anywhere in the path
        filter_spec = filter_by_name(model, ('A',))
    elif frozen == 'class_type_MLP':
        # NOTE: filter based on the class instance
        filter_spec = filter_all_modules_by_type(model, (eqx.nn.MLP,))
    elif frozen == 'class_name_MLP':
        # NOTE: filter based on the string name .__clas__.__name__
        filter_spec = filter_all_modules_by_class_name(model, ('MLP',))

    def loss_fn(model, x, y):
        pred_y = jax.vmap(model)(x)
        return jnp.mean((y - pred_y) ** 2)

    def loss_split(diff_model, static_model, x, y):
        model = eqx.combine(diff_model, static_model)
        return loss_fn(model, x, y)

    @eqx.filter_jit
    def make_step(model, x, y, opt_state):
        if frozen:
            diff_model, static_model = eqx.partition(model, filter_spec)
            grads = eqx.filter_grad(loss_split)(diff_model, static_model, x, y)
        else:
            grads = eqx.filter_grad(loss_fn)(model, x, y)
        updates, opt_state = optim.update(grads, opt_state)
        model = eqx.apply_updates(model, updates)
        return model, opt_state

    original_model = model
    optim = optax.sgd(learning_rate)
    opt_state = optim.init(model)
    for step, (x, y) in zip(range(steps), data_iter):
        model, opt_state = make_step(model, x, y, opt_state)
    if frozen == 'name_bias':
        # NOTE: bias should not change but everything else should
        assert not jnp.allclose(model.A.layers[0].weight, original_model.A.layers[0].weight)
        assert jnp.allclose(model.A.layers[0].bias, original_model.A.layers[0].bias)
        assert jnp.allclose(model.B.layers[0].bias, original_model.B.layers[0].bias)
        assert jnp.allclose(model.C.bias, original_model.C.bias)
    elif frozen == 'name_A':
        # NOTE: everythign in A should be frozen but everything else should change
        assert jnp.allclose(model.A.layers[0].weight, original_model.A.layers[0].weight)
        assert not jnp.allclose(model.B.layers[0].weight, original_model.B.layers[0].weight)
        assert not jnp.allclose(model.B.layers[0].bias, original_model.B.layers[0].bias)
        assert not jnp.allclose(model.C.bias, original_model.C.bias)
    elif frozen == 'class_type_MLP':
        # NOTE: everything in any MLP should be frozen but everything else should change
        assert jnp.allclose(model.A.layers[0].weight, original_model.A.layers[0].weight)
        assert jnp.allclose(model.A.layers[0].bias, original_model.A.layers[0].bias)
        assert jnp.allclose(model.B.layers[0].weight, original_model.B.layers[0].weight)
        assert jnp.allclose(model.B.layers[0].bias, original_model.B.layers[0].bias)
        assert not jnp.allclose(model.C.bias, original_model.C.bias)
        assert not jnp.allclose(model.C.weight, original_model.C.weight)
    elif frozen == 'class_name_MLP':
        assert jnp.allclose(model.A.layers[0].weight, original_model.A.layers[0].weight)
        assert jnp.allclose(model.A.layers[0].bias, original_model.A.layers[0].bias)
        assert jnp.allclose(model.B.layers[0].weight, original_model.B.layers[0].weight)
        assert jnp.allclose(model.B.layers[0].bias, original_model.B.layers[0].bias)
        assert not jnp.allclose(model.C.bias, original_model.C.bias)
        assert not jnp.allclose(model.C.weight, original_model.C.weight)
    elif frozen is None:
        assert not jnp.allclose(model.A.layers[0].weight, original_model.A.layers[0].weight)
        assert not jnp.allclose(model.A.layers[0].bias, original_model.A.layers[0].bias)
        assert not jnp.allclose(model.B.layers[0].weight, original_model.B.layers[0].weight)
        assert not jnp.allclose(model.B.layers[0].bias, original_model.B.layers[0].bias)
        assert not jnp.allclose(model.C.bias, original_model.C.bias)
        assert not jnp.allclose(model.C.weight, original_model.C.weight)


def run_all():
    for k in  ['name_bias', 'name_A', 'class_name_MLP', 'class_type_MLP']:
        print(f'Running with frozen={k} ... ', end='')
        main(frozen=k)
        print('Passed!')

if __name__ == '__main__':
    run_all()
    ```

**Comment by johannahaffner** [comment]
Glad it is working!
